File: src/utils/debug.ts
import chalk from 'chalk';

/**
 * Global debug state - set via --debug flag or DEBUG env var
 */
let debugEnabled = false;

/**
 * Enable debug logging
 */
export function enableDebug(): void {
  debugEnabled = true;
}

/**
 * Check if debug is enabled
 */
export function isDebugEnabled(): boolean {
  return debugEnabled || process.env.DEBUG === 'true' || process.env.DEBUG === '1';
}

/**
 * Recursively filters out vector arrays from objects to make debug logs more readable
 */
function filterVectors(obj: any): any {
  if (obj === null || obj === undefined) {
    return obj;
  }

  // If it's an array with all numbers and length > 10, it's likely a vector
  if (Array.isArray(obj)) {
    if (obj.length > 10 && obj.every(item => typeof item === 'number')) {
      return `[vector with ${obj.length} dimensions]`;
    }
    return obj.map(filterVectors);
  }

  // If it's an object, recursively filter its properties
  if (typeof obj === 'object') {
    const filtered: any = {};
    for (const [key, value] of Object.entries(obj)) {
      // Skip keys named 'vector' or attributes that are large number arrays
      if (key === 'vector' && Array.isArray(value) && value.length > 10) {
        filtered[key] = `[vector with ${value.length} dimensions]`;
      } else {
        filtered[key] = filterVectors(value);
      }
    }
    return filtered;
  }

  return obj;
}

/**
 * Log debug information (only shown when debug is enabled)
 */
export function debugLog(label: string, data: any): void {
  if (isDebugEnabled()) {
    console.log(chalk.gray(`\n[DEBUG] ${label}:`));
    const filtered = filterVectors(data);
    console.log(chalk.gray(typeof filtered === 'string' ? filtered : JSON.stringify(filtered, null, 2)));
  }
}

/**
 * Log raw API request
 */
export function debugRequest(method: string, url: string, payload?: any): void {
  if (isDebugEnabled()) {
    console.log(chalk.gray('\n[DEBUG] API Request:'));
    console.log(chalk.gray(`  Method: ${method}`));
    console.log(chalk.gray(`  URL: ${url}`));
    if (payload) {
      console.log(chalk.gray('  Payload:'));
      console.log(chalk.gray('  ' + JSON.stringify(payload, null, 2).replace(/\n/g, '\n  ')));
    }
  }
}

/**
 * Log raw API response
 */
export function debugResponse(status: number, data: any): void {
  if (isDebugEnabled()) {
    console.log(chalk.gray('\n[DEBUG] API Response:'));
    console.log(chalk.gray(`  Status: ${status}`));
    console.log(chalk.gray('  Data:'));
    console.log(chalk.gray('  ' + JSON.stringify(data, null, 2).replace(/\n/g, '\n  ')));
  }
}

File: src/tpuff/utils/debug.py
"""Debug logging utilities for tpuff CLI."""

import json
import os
from typing import Any

from rich.console import Console

# Console for debug output
console = Console(stderr=True)

# Global debug state - set via --debug flag or DEBUG env var
_debug_enabled = False


def enable_debug() -> None:
    """Enable debug logging."""
    global _debug_enabled
    _debug_enabled = True


def is_debug_enabled() -> bool:
    """Check if debug is enabled."""
    return _debug_enabled or os.environ.get("DEBUG") in ("true", "1")


def filter_vectors(obj: Any) -> Any:
    """Recursively filters out vector arrays from objects to make debug logs more readable."""
    if obj is None:
        return obj

    # If it's a list with all numbers and length > 10, it's likely a vector
    if isinstance(obj, list):
        if len(obj) > 10 and all(isinstance(item, (int, float)) for item in obj):
            return f"[vector with {len(obj)} dimensions]"
        return [filter_vectors(item) for item in obj]

    # If it's a dict, recursively filter its properties
    if isinstance(obj, dict):
        filtered = {}
        for key, value in obj.items():
            # Skip keys named 'vector' or attributes that are large number arrays
            if key == "vector" and isinstance(value, list) and len(value) > 10:
                filtered[key] = f"[vector with {len(value)} dimensions]"
            else:
                filtered[key] = filter_vectors(value)
        return filtered

    return obj


def debug_log(label: str, data: Any) -> None:
    """Log debug information (only shown when debug is enabled)."""
    if is_debug_enabled():
        console.print(f"\n[dim][DEBUG] {label}:[/dim]")
        filtered = filter_vectors(data)
        if isinstance(filtered, str):
            console.print(f"[dim]{filtered}[/dim]")
        else:
            console.print(f"[dim]{json.dumps(filtered, indent=2, default=str)}[/dim]")


def debug_request(method: str, url: str, payload: Any = None) -> None:
    """Log raw API request."""
    if is_debug_enabled():
        console.print("\n[dim][DEBUG] API Request:[/dim]")
        console.print(f"[dim]  Method: {method}[/dim]")
        console.print(f"[dim]  URL: {url}[/dim]")
        if payload:
            console.print("[dim]  Payload:[/dim]")
            payload_str = json.dumps(payload, indent=2, default=str)
            for line in payload_str.split("\n"):
                console.print(f"[dim]  {line}[/dim]")


def debug_response(status: int, data: Any) -> None:
    """Log raw API response."""
    if is_debug_enabled():
        console.print("\n[dim][DEBUG] API Response:[/dim]")
        console.print(f"[dim]  Status: {status}[/dim]")
        console.print("[dim]  Data:[/dim]")
        data_str = json.dumps(data, indent=2, default=str)
        for line in data_str.split("\n"):
            console.print(f"[dim]  {line}[/dim]")
